File: python/player.py
from coordinates import Coordinate
import random


class Player(object):
    def __init__(self):
        self.vision = None
        self.dna = None

    def take_turn(self, dna, vision):
        self.vision = vision
        self.dna = dna
        return self.turn()

    def bit_at(self, position):
        return (self.dna >> position) & 1

    def bit_range(self, start, stop):
        return (self.dna >> start) & ((1 << (stop-start)) - 1)

    def bit_chunk(self, start, length):
        return (self.dna >> start) & ((1 << length) - 1)

    def turn(self):
        return Coordinate(1, 0)

    def vision_at(self, x, y):
        return self.vision[2+y][2+x]

class ForwardPlayer(Player):
    def turn(self):
        return Coordinate(1, 0)

class RandomPlayer(Player):
    def turn(self):
        return Coordinate(1, random.randint(-1, 1))


class LinearCombinationPlayer(Player):
    def __init__(self):
        Player.__init__(self)
        self.coords = [#Coordinate(-1,-1),
                       #Coordinate( 0,-1),
                       Coordinate( 1, 0),
                       Coordinate( 1,-1),
                       #Coordinate(-1, 0),
                       #Coordinate( 0, 0),
                       #Coordinate(-1, 1),
                       #Coordinate( 0, 1),
                       Coordinate( 1, 1)]
        self.n_moves = len(self.coords)

    def turn(self):
        restricted_coords = [c for c in self.coords if self.vision_at(c.x,c.y)>-1]
        restricted_n_moves = len(restricted_coords)
        s = 0
        for i in range(25):
            s += self.bit_range(2*i,2*i+2)*self.vision_at(int(i/5)-2, i%5-2)
        return restricted_coords[s%restricted_n_moves]

class ColorScorePlayer(Player):
    def __init__(self):
        Player.__init__(self)
        self.coords = [#Coordinate(-1,-1),
                       #Coordinate( 0,-1),
                       Coordinate( 1, 0),
                       Coordinate( 1,-1),
                       #Coordinate(-1, 0),
                       #Coordinate( 0, 0),
                       #Coordinate(-1, 1),
                       #Coordinate( 0, 1),
                       Coordinate( 1, 1)]
        self.n_moves = len(self.coords)

    def turn(self):
        max_score = max([self.bit_chunk(6*self.vision_at(c.x, c.y), 6) for c in self.coords if self.vision_at(c.x, c.y)>=0])
        restricted_coords = [c for c in self.coords if self.vision_at(c.x, c.y)>=0 and self.bit_chunk(6*self.vision_at(c.x,c.y), 6) == max_score]

        return random.choice(restricted_coords)



PLAYER_TYPE = ColorScorePlayer

File: ruby/player.rb
require_relative 'vector2d'

class Player
    def initialize(rng)
        @rng = rng
    end

    def take_turn(genome, vision)
        @genome = genome
        @vision = vision
        turn
    end

    def turn
        Vector2D.new(1,0)
    end

    def bit_at(i)
        (@genome >> i) & 1
    end

    def bit_range(start, stop)
        (@genome >> start) & ((1 << (stop-start)) - 1)
    end

    def bit_chunk(start, length)
        (@genome >> start) & ((1 << length) - 1)
    end

    def vision_at(x, y)
        @vision[x+2][y+2]
    end
end

class RandomPlayer < Player
    def turn
        Vector2D.new(1,@rng.rand(3)-1)
    end
end

class LemmingPlayer < Player
    def initialize(rng)
        super(rng)
        @coords = [Vector2D.new(-1,-1),
                   Vector2D.new(-1, 0),
                   Vector2D.new(-1, 1),
                   Vector2D.new( 1, 0)]
    end

    def turn
        @coords.sample(random: @rng)
    end
end

class ColorScorePlayer < Player
    def initialize(rng)
        super(rng)
        @coords = [Vector2D.new( 1,-1),
                   Vector2D.new( 1, 0),
                   Vector2D.new( 1, 1)]
    end

    def vision_at(vec2d)
        @vision[vec2d.x+2][vec2d.y+2]
    end

    def turn
        max_score = @coords.map { |c|
            color = vision_at(c)
            color < 0 ? -1 : bit_chunk(6*color, 6)
        }.max

        restricted_coords = @coords.select { |c|
            color = vision_at(c)
            color >= 0 && bit_chunk(6*color, 6) == max_score
        }

        restricted_coords.sample(random: @rng)
    end
end

PLAYER = ColorScorePlayer